polymetis/polymetis/python/polysim/envs/experimental/bullet_manipulator.py

```diff
@@ -0,0 +1,164 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import numpy as np
+import hydra
+from omegaconf import OmegaConf
+from omegaconf.dictconfig import DictConfig
+import pybullet
+from pybullet_utils.bullet_client import BulletClient
+
+import polymetis_pb2
+import polysim
+from polymetis.utils.data_dir import get_full_path_to_urdf
+
+
+class BulletManipulator:
+    def __init__(
+        self,
+        cfg: DictConfig,
+        gui: bool,
+        gravity: float = 9.81,
+    ):
+        self.cfg = cfg
+
+        # Initialize PyBullet simulation
+        if gui:
+            self.sim = BulletClient(connection_mode=pybullet.GUI)
+        else:
+            self.sim = BulletClient(connection_mode=pybullet.DIRECT)
+
+        urdf_path = get_full_path_to_urdf(self.cfg.robot_description_path)
+        self.robot_id = self.sim.loadURDF(
+            urdf_path,
+            basePosition=[0.0, 0.0, 0.0],
+            useFixedBase=True,
+            flags=pybullet.URDF_USE_INERTIA_FROM_FILE,
+        )
+
+        for i in range(7):
+            self.sim.resetJointState(
+                bodyUniqueId=self.robot_id,
+                jointIndex=self.cfg.controlled_joints[i],
+                targetValue=self.cfg.rest_pose[i],
+                targetVelocity=0,
+            )
+
+        # Initialize states
+        self.arm_state = polymetis_pb2.RobotState()
+        self.arm_state.prev_joint_torques_computed[:] = np.zeros(7)
+        self.arm_state.prev_joint_torques_computed_safened[:] = np.zeros(7)
+        self.arm_state.motor_torques_measured[:] = np.zeros(7)
+        self.arm_state.motor_torques_external[:] = np.zeros(7)
+
+        self.arm_state.prev_command_successful = True
+        self.arm_state.error_code = 0
+
+        self.gripper_state = polymetis_pb2.GripperState()
+
+    def get_arm_state(self) -> polymetis_pb2.RobotState:
+        # Timestamp
+        self.arm_state.timestamp.GetCurrentTime()
+
+        # Joint pos & vel
+        joint_cur_states = self.sim.getJointStates(
+            self.robot_id, self.cfg.controlled_joints
+        )
+        self.arm_state.joint_positions[:] = [joint_cur_states[i][0] for i in range(7)]
+        self.arm_state.joint_velocities[:] = [joint_cur_states[i][1] for i in range(7)]
+
+        return self.arm_state
+
+    def get_gripper_state(self) -> polymetis_pb2.GripperState:
+        joint_cur_states = self.sim.getJointStates(
```

Comment:
Comparing this to get_arm_state, does gripper_state not need a timestamp?

Reply:
Oops